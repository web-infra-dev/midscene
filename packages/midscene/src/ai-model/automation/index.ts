import assert from 'node:assert';
import type { PlanningAIResponse, PlanningAction, UIContext } from '@/types';
import {
  AIActionType,
  type AIArgs,
  callAiFn,
  transformUserMessages,
} from '../common';
import { describeUserPage } from '../prompt/util';
import { systemPromptToTaskPlanning } from './planning';

export async function plan(
  userPrompt: string,
  opts: {
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
  useModel?: 'coze' | 'openAI',
): Promise<{ plans: PlanningAction[] }> {
  const { callAI, context } = opts || {};
  const { screenshotBase64 } = context;
  const { description: pageDescription } = await describeUserPage(context);
  let planFromAI: PlanningAIResponse | undefined;

  const systemPrompt = systemPromptToTaskPlanning();
  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: `
            pageDescription:\n 
            ${pageDescription}
            \n
            Here is the description of the task. Just go ahead:
            =====================================
            ${userPrompt}
            =====================================
          `,
        },
      ]),
    },
  ];

  if (callAI) {
    planFromAI = await callAI({
      msgs,
      AIActionType: AIActionType.PLAN,
      useModel,
    });
  } else {
    planFromAI = await callAiFn({
      msgs,
      AIActionType: AIActionType.PLAN,
      useModel,
    });
  }

  const actions = planFromAI?.actions || [];

  assert(planFromAI, "can't get planFromAI");
  assert(actions && actions.length > 0, 'no actions in ai plan');

  if (planFromAI.error) {
    throw new Error(planFromAI.error);
  }

  // actions.forEach((task) => {
  //   if (task.type === 'Error') {
  //     throw new Error(task.thought);
  //   }
  // });

  return { plans: actions };
}
