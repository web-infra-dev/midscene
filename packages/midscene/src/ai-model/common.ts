import type OpenAI from 'openai';
import type {
  ChatCompletionContentPart,
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
} from 'openai/resources';
import {
  COZE_AI_ACTION_BOT_ID,
  COZE_AI_ASSERT_BOT_ID,
  COZE_EXTRACT_INFO_BOT_ID,
  COZE_INSPECT_ELEMENT_BOT_ID,
  callCozeAi,
  transfromOpenAiArgsToCoze,
  useCozeModel,
} from './coze';
import {
  MIDSCENE_MODEL_TEXT_ONLY,
  callToGetJSONObject,
  useOpenAIModel,
} from './openai';

export type AIArgs = [
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
];

export enum AIActionType {
  ASSERT = 0,
  INSPECT_ELEMENT = 1,
  EXTRACT_DATA = 2,
  PLAN = 3,
}

export async function callAiFn<T>(options: {
  msgs: AIArgs;
  AIActionType: AIActionType;
  useModel?: 'openAI' | 'coze';
}) {
  const { useModel, msgs, AIActionType: AIActionTypeValue } = options;
  if (useOpenAIModel(useModel)) {
    const parseResult = await callToGetJSONObject<T>(msgs, AIActionTypeValue);
    return parseResult;
  }

  if (useCozeModel(useModel)) {
    let botId = '';
    switch (AIActionTypeValue) {
      case AIActionType.ASSERT:
        botId = COZE_AI_ASSERT_BOT_ID;
        break;
      case AIActionType.EXTRACT_DATA:
        botId = COZE_EXTRACT_INFO_BOT_ID;
        break;
      case AIActionType.INSPECT_ELEMENT:
        botId = COZE_INSPECT_ELEMENT_BOT_ID;
        break;
      default:
        botId = COZE_AI_ACTION_BOT_ID;
    }
    const cozeMsg = transfromOpenAiArgsToCoze(msgs[1]);
    const parseResult = await callCozeAi<T>({
      ...cozeMsg,
      botId,
    });
    return parseResult;
  }

  throw Error(
    'Cannot find Coze or OpenAI config. You should set at least one of them.',
  );
}

export function transformUserMessages(msgs: ChatCompletionContentPart[]) {
  const textOnly = Boolean(process.env[MIDSCENE_MODEL_TEXT_ONLY]);
  if (!textOnly) return msgs;

  return msgs.reduce((res, msg) => {
    if (msg.type === 'text') {
      res += msg.text;
    }
    return res;
  }, '');
}
