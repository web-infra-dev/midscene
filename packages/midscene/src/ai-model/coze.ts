import type { BaseElement, UIContext } from '@/types';
import fetch from 'node-fetch';
import { multiDescription } from './prompt/element_inspector';
import { describeUserPage } from './prompt/util';

const INSPECT_ELEMENT_BOT_ID = '7390985487806775304';

export async function CozeAiInspectElement<
  ElementType extends BaseElement = BaseElement,
>(msg: {
  findElementDescription: string;
  context: UIContext<ElementType>;
  multi?: boolean;
}) {
  const { multi = false, findElementDescription, context } = msg;
  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);
  const query = JSON.stringify({
    description: findElementDescription,
    multi: multiDescription(multi),
    elementDescription: description,
  });
  const parseResult = await fetch('https://api.coze.com/open_api/v2/chat', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${process.env.COZE_BOT_TOKEN}`,
      'Content-Type': 'application/json',
      Accept: '*/*',
      Host: 'api.coze.com',
      Connection: 'keep-alive',
    },
    body: JSON.stringify({
      conversation_id: '123',
      bot_id: INSPECT_ELEMENT_BOT_ID,
      user: '29032201862555',
      query,
      meta_data: {
        img: [
          {
            url: screenshotBase64,
          },
        ],
      },
      stream: false,
    }),
  })
    .then((response) => {
      if (!response.ok) {
        console.log('response', response);
        throw new Error('Network response was not ok');
      }
      return response.json();
    })
    .then((data) => {
      const parseContent = data?.messages[0]?.content;
      if (parseContent) {
        return JSON.parse(parseContent);
      }
      return data;
    })
    .catch((error) => {
      console.error('Fetch error:', error);
    });
  return {
    parseResult,
    elementById,
  };
}
