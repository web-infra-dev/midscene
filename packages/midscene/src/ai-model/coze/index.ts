import assert from 'node:assert';
import type { ChatCompletionUserMessageParam } from 'openai/resources';

export const COZE_INSPECT_ELEMENT_BOT_ID =
  process.env.COZE_INSPECT_ELEMENT_BOT_ID || '';
export const COZE_AI_ACTION_BOT_ID = process.env.COZE_AI_ACTION_BOT_ID || '';
export const COZE_AI_ASSERT_BOT_ID = process.env.COZE_AI_ASSERT_BOT_ID || '';
export const COZE_EXTRACT_INFO_BOT_ID =
  process.env.COZE_EXTRACT_INFO_BOT_ID || '';

export const COZE_BOT_TOKEN = 'COZE_BOT_TOKEN';

export function preferCozeModel(preferVendor?: 'coze' | 'openAI') {
  if (preferVendor && preferVendor !== 'coze') return false;
  return (
    process.env[COZE_BOT_TOKEN] &&
    process.env.COZE_INSPECT_ELEMENT_BOT_ID &&
    process.env.COZE_AI_ACTION_BOT_ID &&
    process.env.COZE_AI_ASSERT_BOT_ID &&
    process.env.COZE_EXTRACT_INFO_BOT_ID
  );
}

export async function callCozeAi<T>(options: {
  query: string;
  imgs: Array<string>;
  botId: string;
}): Promise<T> {
  const { query, imgs, botId } = options;
  const completion = await fetch('https://api.coze.com/open_api/v2/chat', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${process.env[COZE_BOT_TOKEN]}`,
      'Content-Type': 'application/json',
      Accept: '*/*',
      Host: 'api.coze.com',
      Connection: 'keep-alive',
    },
    body: JSON.stringify({
      conversation_id: '123',
      bot_id: botId,
      user: '29032201862555',
      query,
      meta_data: {
        img: imgs.map((imgPath) => {
          return {
            url: imgPath,
          };
        }),
      },
      stream: false,
    }),
  });
  if (!completion.ok) {
    console.error('CozeAI reponse error', completion);
    throw new Error('Network response was not ok');
  }

  const aiResponse = await completion.json();
  if (aiResponse.code !== 0) {
    console.error('CozeAI error response', aiResponse.msg);
    throw new Error(`CozeAI error response ${aiResponse.msg}`);
  }

  if (!aiResponse?.messages || !aiResponse?.messages[0]?.content) {
    console.error('aiResponse', aiResponse);
    throw new Error('aiResponse is undefined', aiResponse);
  }
  const parseContent = aiResponse?.messages[0]?.content;
  assert(parseContent, 'empty content');
  try {
    return JSON.parse(parseContent);
  } catch (err) {
    console.error("can't parse coze content", aiResponse, err);
    throw Error("can't parse coze content");
  }
}

export function transformOpenAiArgsToCoze(msg: ChatCompletionUserMessageParam) {
  if (msg.role !== 'user') throw Error(`can't transform ${msg} to coze args`);
  // const query = '';
  // const imgs = msg.content
  if (typeof msg.content === 'string') {
    return {
      query: msg.content,
      imgs: [],
    };
  }

  return {
    query: msg.content.reduce((res, next) => {
      if (next.type === 'text') {
        res += `\n${next.text}`;
      }
      return res;
    }, ''),
    imgs: msg.content.reduce(
      (res, next) => {
        if (next.type === 'image_url') {
          res.push(next.image_url.url);
        }
        return res;
      },
      [] as Array<string>,
    ),
  };
}
