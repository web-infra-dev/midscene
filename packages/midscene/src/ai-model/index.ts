export { callToGetJSONObject } from './openai';
export { systemPromptToFindElement } from './prompt/element_inspector';
export { describeUserPage } from './prompt/util';

export type { ChatCompletionMessageParam } from 'openai/resources';

export {
  AiInspectElement,
  AiExtractElementInfo,
  AiAssert,
  transformElementPositionToId,
} from './inspect';

export { plan } from './automation';
export { callAiFn } from './common';
