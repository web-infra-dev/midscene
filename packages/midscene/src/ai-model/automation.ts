import assert from 'node:assert';
import type { AIUsageInfo, PlanningAIResponse, UIContext } from '@/types';
import { PromptTemplate } from '@langchain/core/prompts';
import { AIActionType, type AIArgs, callAiFn } from './common';
import {
  automationUserPrompt,
  systemPromptToTaskPlanning,
  taskBackgroundContext,
} from './prompt/planning';
import { describeUserPage } from './prompt/util';

export async function plan(
  userPrompt: string,
  opts: {
    whatHaveDone?: string;
    originalPrompt?: string;
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
): Promise<PlanningAIResponse> {
  const { callAI, context } = opts || {};
  const { screenshotBase64, screenshotBase64WithElementMarker } = context;
  const { description: pageDescription, elementByPosition } =
    await describeUserPage(context);

  const systemPrompt = await systemPromptToTaskPlanning();
  const userInstructionPrompt = await automationUserPrompt.format({
    pageDescription,
    userPrompt,
    taskBackgroundContext: taskBackgroundContext(
      opts.originalPrompt,
      opts.whatHaveDone,
    ),
  });

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64WithElementMarker || screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: userInstructionPrompt,
        },
      ],
    },
  ];

  const call = callAI || callAiFn;
  const { content, usage } = await call(msgs, AIActionType.PLAN);
  const planFromAI = content;

  const actions = planFromAI?.actions || [];
  assert(planFromAI, "can't get plans from AI");
  assert(
    actions.length > 0,
    `Failed to plan actions with context: ${planFromAI.error}`,
  );

  return planFromAI;
}
