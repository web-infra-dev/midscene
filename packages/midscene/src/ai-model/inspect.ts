import assert from 'node:assert';
import type {
  AIAssertionResponse,
  AIElementParseResponse,
  AISectionParseResponse,
  BaseElement,
  UIContext,
} from '@/types';
import type {
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
} from 'openai/resources';
import { AIActionType, callAiFn, transformUserMessages } from './common';
import {
  multiDescription,
  systemPromptToFindElement,
} from './prompt/element_inspector';
import {
  describeUserPage,
  systemPromptToAssert,
  systemPromptToExtract,
} from './prompt/util';

export type AIArgs = [
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
];

export async function AiInspectElement<
  ElementType extends BaseElement = BaseElement,
>(options: {
  context: UIContext<ElementType>;
  multi: boolean;
  findElementDescription: string;
  callAI?: typeof callAiFn<AIElementParseResponse>;
  useModel?: 'coze' | 'openAI';
}) {
  const { context, multi, findElementDescription, callAI, useModel } = options;
  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);

  const systemPrompt = systemPromptToFindElement();

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
          },
        },
        {
          type: 'text',
          text: `
    pageDescription: \n
    ${description}

    Here is the description of the findElement. Just go ahead:
    =====================================
    ${JSON.stringify({
      description: findElementDescription,
      multi: multiDescription(multi),
    })}
    =====================================
  `,
        },
      ]),
    },
  ];

  if (callAI) {
    const parseResult = await callAI({
      msgs,
      AIActionType: AIActionType.INSPECT_ELEMENT,
      useModel,
    });
    return {
      parseResult,
      elementById,
    };
  }
  const inspectElement = await callAiFn<AIElementParseResponse>({
    msgs,
    AIActionType: AIActionType.INSPECT_ELEMENT,
    useModel,
  });

  return {
    parseResult: inspectElement,
    elementById,
  };
}

export async function AiExtractElementInfo<
  T,
  ElementType extends BaseElement = BaseElement,
>(options: {
  dataQuery: string | Record<string, string>;
  context: UIContext<ElementType>;
  useModel?: 'coze' | 'openAI';
}) {
  const { dataQuery, context, useModel } = options;
  const systemPrompt = systemPromptToExtract();

  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
          },
        },
        {
          type: 'text',
          text: `
pageDescription: ${description}

Use your extract_data_from_UI skill to find the following data, placing it in the \`data\` field
DATA_DEMAND start:
=====================================
${
  typeof dataQuery === 'object'
    ? `return in key-value style object, keys are ${Object.keys(dataQuery).join(',')}`
    : ''
};
${typeof dataQuery === 'string' ? dataQuery : JSON.stringify(dataQuery, null, 2)}
=====================================
DATA_DEMAND ends.
          `,
        },
      ]),
    },
  ];

  const parseResult = await callAiFn<AISectionParseResponse<T>>({
    msgs,
    useModel,
    AIActionType: AIActionType.EXTRACT_DATA,
  });
  return {
    parseResult,
    elementById,
  };
}

export async function AiAssert<
  ElementType extends BaseElement = BaseElement,
>(options: {
  assertion: string;
  context: UIContext<ElementType>;
  useModel?: 'coze' | 'openAI';
}) {
  const { assertion, context, useModel } = options;

  assert(assertion, 'assertion should be a string');

  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);
  const systemPrompt = systemPromptToAssert();

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
          },
        },
        {
          type: 'text',
          text: `
    pageDescription: \n
    ${description}
    Here is the description of the assertion. Just go ahead:
    =====================================
    ${assertion}
    =====================================
  `,
        },
      ]),
    },
  ];

  const assertResult = await callAiFn<AIAssertionResponse>({
    msgs,
    AIActionType: AIActionType.ASSERT,
    useModel,
  });
  return assertResult;
}
export { callAiFn };
