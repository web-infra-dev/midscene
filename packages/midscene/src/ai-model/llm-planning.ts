import assert from 'node:assert';
import { MIDSCENE_USE_QWEN_VL, getAIConfigInBoolean } from '@/env';
import type {
  PlanningAIResponse,
  PlanningLocateParam,
  UIContext,
} from '@/types';
import { AIActionType, type AIArgs, callAiFn } from './common';
import {
  automationUserPrompt,
  generateTaskBackgroundContext,
  systemPromptToTaskPlanning,
} from './prompt/llm-planning';
import { describeUserPage } from './prompt/util';

// transform the param of locate from qwen mode
export function fillLocateParam(locate: PlanningLocateParam) {
  if (locate?.bbox_2d && !locate?.bbox) {
    locate.bbox = locate.bbox_2d;
    // biome-ignore lint/performance/noDelete: <explanation>
    delete locate.bbox_2d;
  }

  const defaultBboxSize = 10;
  if (locate?.bbox) {
    locate.bbox[0] = Math.round(locate.bbox[0]);
    locate.bbox[1] = Math.round(locate.bbox[1]);
    locate.bbox[2] =
      typeof locate.bbox[2] === 'number'
        ? Math.round(locate.bbox[2])
        : Math.round(locate.bbox[0] + defaultBboxSize);
    locate.bbox[3] =
      typeof locate.bbox[3] === 'number'
        ? Math.round(locate.bbox[3])
        : Math.round(locate.bbox[1] + defaultBboxSize);
  }

  return locate;
}

export async function plan(
  userInstruction: string,
  opts: {
    log?: string;
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
): Promise<PlanningAIResponse> {
  const { callAI, context } = opts || {};
  const { screenshotBase64, screenshotBase64WithElementMarker, size } = context;
  const { description: pageDescription } = await describeUserPage(context);

  const systemPrompt = await systemPromptToTaskPlanning();
  const taskBackgroundContextText = generateTaskBackgroundContext(
    userInstruction,
    opts.log,
  );
  const userInstructionPrompt = await automationUserPrompt().format({
    pageDescription,
    taskBackgroundContext: taskBackgroundContextText,
  });

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64WithElementMarker || screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: userInstructionPrompt,
        },
      ],
    },
  ];

  const call = callAI || callAiFn;
  const { content, usage } = await call(msgs, AIActionType.PLAN);
  const rawResponse = JSON.stringify(content, undefined, 2);
  const planFromAI = content;
  const actions =
    (planFromAI.action ? [planFromAI.action] : planFromAI.actions) || [];
  const returnValue: PlanningAIResponse = {
    ...planFromAI,
    actions,
    rawResponse,
    usage,
  };

  if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    actions.forEach((action) => {
      if (action.locate) {
        action.locate = fillLocateParam(action.locate);
      }
    });
  }

  assert(planFromAI, "can't get plans from AI");
  assert(
    actions.length > 0 || returnValue.finish,
    `Failed to plan actions: ${planFromAI.error || '(no error details)'}`,
  );

  return returnValue;
}
