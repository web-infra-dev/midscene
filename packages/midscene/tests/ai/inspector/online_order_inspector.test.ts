import path from 'node:path';
import { AiInspectElement } from '@/ai-model';
import { expect, test } from 'vitest';
import {
  getPageTestData,
  repeat,
  runTestCases,
  writeFileSyncWithDir,
} from './util';

const testCases = [
  {
    description: 'Top left menu bar icon',
    multi: false,
  },
  {
    description: 'Toggle language text button(Could be：中文、english text)',
    multi: false,
  },
  {
    description: 'Top right shopping cart',
    multi: false,
  },
  {
    description: 'The price number on the right of the drink picture',
    multi: true,
  },
  {
    description: '选择规格按钮',
    multi: true,
  },
  {
    description: 'Bottom right Customer service button',
    multi: false,
  },
];

const repeatTime = process.env.GITHUB_ACTIONS ? 1 : 5;

repeat(repeatTime, (repeatIndex) => {
  test(
    'xicha: inspect element',
    async () => {
      const { context } = await getPageTestData(
        path.join(__dirname, './test-data/online_order'),
      );

      const { aiResponse, filterUnstableResult } = await runTestCases(
        testCases,
        context,
        async (testCase) => {
          const { parseResult } = await AiInspectElement({
            context,
            multi: testCase.multi,
            findElementDescription: testCase.description,
          });
          return parseResult;
        },
      );
      writeFileSyncWithDir(
        path.join(
          __dirname,
          `__ai_responses__/online_order-inspector-element-${repeatIndex}.json`,
        ),
        JSON.stringify(aiResponse, null, 2),
        { encoding: 'utf-8' },
      );
      expect(filterUnstableResult).toMatchFileSnapshot(
        './__snapshots__/online_order_inspector.test.ts.snap',
      );
    },
    {
      timeout: 90 * 1000,
    },
  );
});
