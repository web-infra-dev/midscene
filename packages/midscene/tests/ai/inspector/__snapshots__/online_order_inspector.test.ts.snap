[
  {
    "elements": [
      {
        "id": "6ad26dfdca",
        "indexId": "1",
      },
    ],
    "error": [],
    "prompt": "Top left menu bar icon",
  },
  {
    "elements": [
      {
        "id": "e666c51ff8",
        "indexId": "2",
      },
    ],
    "error": [],
    "prompt": "Toggle language text button(Could be：中文、english text)",
  },
  {
    "elements": [
      {
        "id": "f775c69cb4",
        "indexId": "4",
      },
    ],
    "error": [],
    "prompt": "Top right shopping cart",
  },
  {
    "elements": [
      {
        "id": "14103376fb",
        "indexId": "25",
      },
      {
        "id": "0250e12e67",
        "indexId": "31",
      },
    ],
    "error": [],
    "prompt": "The price number on the right of the drink picture",
  },
  {
    "elements": [
      {
        "id": "fadfa10182",
        "indexId": "26",
      },
      {
        "id": "001617dbf8",
        "indexId": "32",
      },
    ],
    "error": [],
    "prompt": "选择规格按钮",
  },
  {
    "elements": [
      {
        "id": "cad3004a2d",
        "indexId": "33",
      },
    ],
    "error": [],
    "prompt": "Bottom right Customer service button",
  },
]